src/System.Threading/tests/ReaderWriterLockSlimTests.cs

```diff
@@ -448,5 +448,67 @@ public static void ReleaseReadersWhenWaitingWriterTimesOut()
                 writeWaiterThread.Join();
             }
         }
+
+        [Fact]
+        public static void DontReleaseWaitingReadersWhenThereAreWaitingWriters()
+        {
+            using(var rwls = new ReaderWriterLockSlim())
+            {
+                rwls.EnterUpgradeableReadLock();
+                rwls.EnterWriteLock();
+                // Typical order of execution: 0
+
+                // Add a waiting writer
+                var threads = new Thread[2];
+                using(var beforeEnterWriteLock = new ManualResetEvent(false))
+                {
+                    var thread =
+                        new Thread(() =>
+                        {
+                            beforeEnterWriteLock.Set();
+                            rwls.EnterWriteLock();
+                            // Typical order of execution: 3
+                            rwls.ExitWriteLock();
+                        });
+                    thread.IsBackground = true;
+                    thread.Start();
+                    threads[0] = thread;
+                    beforeEnterWriteLock.WaitOne();
+                }
+
+                // Add a waiting reader
+                using(var beforeEnterReadLock = new ManualResetEvent(false))
+                {
+                    var thread =
+                        new Thread(() =>
+                        {
+                            beforeEnterReadLock.Set();
+                            rwls.EnterReadLock();
+                            // Typical order of execution: 4
+                            rwls.ExitReadLock();
+                        });
+                    thread.IsBackground = true;
+                    thread.Start();
+                    threads[1] = thread;
+                    beforeEnterReadLock.WaitOne();
+                }
+
+                // Wait for the background threads to block waiting for their locks
+                Thread.Sleep(1000);
```

Comment:
The test will work with any order of execution, just may not frequently hit the case being tested for without the sleep. I have marked this and the test above as [OuterLoop].